src/main/java/net/minecraftforge/event/entity/SpawnPlacementRegisterEvent.java

```diff
@@ -0,0 +1,207 @@
+import net.minecraftforge.eventbus.api.Cancelable;
+import net.minecraftforge.eventbus.api.Event;
+import net.minecraftforge.eventbus.api.EventPriority;
+import net.minecraftforge.fml.event.IModBusEvent;
+
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * This event allows each {@link EntityType} to have a {@link SpawnPlacements.SpawnPredicate} registered or modified.
+ * Spawn Predicates are checked whenever an {@link Entity} of the given {@link EntityType} spawns in the world naturally.
+ *
+ * If registering your own entity's spawn placements, you should use {@link SpawnPlacementRegisterEvent#register(EntityType, SpawnPlacements.Type, Heightmap.Types, SpawnPlacements.SpawnPredicate, Operation)}
+ * So that you ensure that your entity has a heightmap type and placement type registered.
+ *
+ * If modifying vanilla or another mod's spawn placements, you can use three operations:
+ *  REPLACE: checked first, the last mod to replace the predicate wipes out all other predicates. Listen with a low {@link EventPriority} if you need to do this.
+ *  OR: checked second, only one of these predicates must pass along with the original predicate
+ *  AND: checked third, these predicates must all pass along with the original predicate
+ *
+ * <p>
+ * This event is not {@linkplain Cancelable cancellable} and does not {@linkplain Event.HasResult have a result}.
+ * <p>
+ *
+ *  Fired on the Mod bus {@link IModBusEvent}.<br>
+ */
+public class SpawnPlacementRegisterEvent extends Event implements IModBusEvent
+{
+    private final Map<EntityType<?>, MergedSpawnPredicate> map;
+
+    @ApiStatus.Internal
+    public SpawnPlacementRegisterEvent(Map<EntityType<?>, MergedSpawnPredicate> map)
+    {
+        this.map = map;
+    }
+
+    /**
+     * Register an optional spawn placement {@code predicate} for a given {@code entityType}
+     */
+    public void register(EntityType<?> entityType, SpawnPlacements.SpawnPredicate<Entity> predicate)
+    {
+        register(entityType, null, null, predicate, Operation.OR);
+    }
+
+    /**
+     * Register a {@code predicate} for a given {@code entityType} with a given {@code operation} for handling
+     */
+    public void register(EntityType<?> entityType, SpawnPlacements.SpawnPredicate<Entity> predicate, Operation operation)
+    {
+        register(entityType, null, null, predicate, operation);
+    }
+
+    /**
+     * Register a {@code predicate} for a given {@code entityType} and {@code operation}
+     * With the option of changing the {@code placementType} and {@code heightmap}. These are only applied if {@link Operation#REPLACE} is used.
+     * Use {@code null} for the placement or heightmap to leave them as is (which should be done in almost every case)
+     */
+    public void register(EntityType<?> entityType, @Nullable SpawnPlacements.Type placementType, @Nullable Heightmap.Types heightmap, SpawnPlacements.SpawnPredicate<Entity> predicate, Operation operation)
```

Comment:
You can genericize these methods so that the spawn predicate can use the correct instance of the entity without casting. Additionally, should this be only limited to mobs, or all entities (only used for mobs currently).

Reply:
gonna leave the type bound as entity as that's the type bound for SpawnPlacements